core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -335,16 +338,23 @@ namespace kagome::consensus::grandpa {
           break;
         }
 
-        const auto &maybe_finalized = last_round_state.finalized;
+        const auto &maybe_last_finalized = last_round_state.finalized;
+        const auto &maybe_curr_finalized = cur_round_state_.finalized;
 
-        // we should send primary if last round's state contains finalized
-        // block and last round estimate is better than finalized block
-        bool should_send_primary =
-            maybe_finalized
-                .map([&maybe_estimate](const BlockInfo &finalized) {
-                  return maybe_estimate->block_number > finalized.block_number;
-                })
-                .value_or(false);
+        bool should_send_primary = false;
+        // We should send primary if last round estimate was not finalized in
+        // last round
+        if (maybe_last_finalized
+            and maybe_estimate->block_number
+                    > maybe_last_finalized->block_number) {
+          should_send_primary = true;
+        }
+        // Or if last round estimate was not finalized in current round
+        else {
+          should_send_primary = maybe_curr_finalized
```

Comment:
Reposting from gitlab:

This logic seem to have multiple problems. if nothing has been finalized in the last round then last round estimate hasn't been finalized hence primary should be sent. (so why not ` if maybe_finalized == false then shoud_send_primary = true`)
Also if maybe_estimate->block_number > maybe_last_finalized->block_number that is false it goes to else which then if maybe_curr_finalized doesn't exsits makes should_send_primary becomes false. This is probably what we wants but it is very confusing to hide the logic in such an convoluted riddlish fashion.